### PR TITLE
Refactor the xlinear model for batch prediction

### PR DESCRIPTION
**Pull request by hallogameboy**
*Issue #, if available:*

*Description of changes:*

Refactor the xlinear model to support the batch prediction.
Add a test for validating the correctness of batch prediction results.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/xmc/xlinear/model.py [hunk 1]
```diff
@@ -427,6 +427,7 @@ def predict(
         X,
         pred_params=None,
         selected_outputs_csr=None,
+        batch_size=None,
```
Comment (jiong-zhang) [inline pecos/xmc/xlinear/model.py]: I would suggest to include `batch_size` into `kwargs`.

File: pecos/xmc/xlinear/model.py [hunk 2]
```diff
@@ -437,37 +438,56 @@ def predict(
...
+                    Y_pred = self.model.predict_on_selected_outputs(
+                        X, selected_outputs_csr, pred_params=pred_params.hlm_args, **kwargs
+                    )
+                else:
+                    raise TypeError("type(pred_kwargs) is not supported")
+        elif isinstance(batch_size, int):
+            Ys = []
+            for i in range(0, X.shape[0], batch_size):
+                Ye = self.predict(
+                    X=X[i : i + batch_size, :],
```
Comment (jiong-zhang) [inline pecos/xmc/xlinear/model.py]: `X` is not a keyword argument here.

File: pecos/xmc/xlinear/model.py [hunk 3]
```diff
@@ -437,37 +438,56 @@ def predict(
...
                 scores is returned.
         """
-        if selected_outputs_csr is None:
-            if pred_params is None:
-                Y_pred = self.model.predict(X, pred_params=None, **kwargs)
-            elif isinstance(pred_params, self.PredParams):
-                Y_pred = self.model.predict(X, pred_params=pred_params.hlm_args, **kwargs)
+        if batch_size is None:
+            if selected_outputs_csr is None:
+                if pred_params is None:
```
Comment (jiong-zhang) [inline pecos/xmc/xlinear/model.py]: Can we change this (and line466) to `Y_pred = self.model.predict(X, pred_params=None if pred_params is None else pred_params.hlm_args, **kwargs)`

**Review by jiong-zhang — COMMENTED** [review]
In addition to the changes in `pecos/xmc/xlinear/model.py`, we should also remove the loop in `pecos/xmc/xlinear/predict.py`.

File: pecos/xmc/xlinear/model.py (hunk 1 above)
Comment (hallogameboy) [reply pecos/xmc/xlinear/model.py]: Thanks for the suggestion. I will change it in the amendment.

File: pecos/xmc/xlinear/model.py (hunk 2 above)
Comment (hallogameboy) [reply pecos/xmc/xlinear/model.py]: Thanks for the catch. I will fix in the amendment.

File: pecos/xmc/xlinear/model.py (hunk 3 above)
Comment (hallogameboy) [reply pecos/xmc/xlinear/model.py]: Thanks for the suggestion. I think this original style was aiming to catch and raise the error for invalid types of pred_params, but the suggested style indeed has better readability. I will change it to the suggested style and add a check for the instance type ahead.

File: pecos/xmc/xlinear/model.py [hunk 4]
```diff
@@ -444,30 +444,49 @@ def predict(
...
         """
-        if selected_outputs_csr is None:
-            if pred_params is None:
-                Y_pred = self.model.predict(X, pred_params=None, **kwargs)
-            elif isinstance(pred_params, self.PredParams):
-                Y_pred = self.model.predict(X, pred_params=pred_params.hlm_args, **kwargs)
+        if (pred_params is not None) and (not isinstance(pred_params, self.PredParams)):
+            raise TypeError("type(pred_kwargs) is not supported")
+
+        if "batch_size" not in kwargs:
```
Comment (jiong-zhang) [inline pecos/xmc/xlinear/model.py]: Now `kwargs = {“batch_size”: None}` will raise error. You can use `batch_size=kwargs.pop("batch_size", None)` at beggining.

Comment (hallogameboy) [reply pecos/xmc/xlinear/model.py]: Thanks for the catch. Fixed this in the amendment.

**Review by jiong-zhang — APPROVED** [review]

**Pull request merged**